### PR TITLE
PK: RSA decryption

### LINKED ISSUE #5160: PK: RSA decryption
In `library/pk_wrap.c`, provide an implementation of `rsa_decrypt_wrap` using `psa_asymmetric_decrypt()` instead of `mbedtls_rsa_pkcs1_decrypt()` when `MBEDTLS_USE_PSA_CRYPTO` is enabled.

This will require creating a temporary PSA private key with appropriate permissions. An example of something similar, but with a public key, can be found in `ecdsa_verify_wrap()` in the same file (note: there are two definitions of `ecdsa_very_wrap()`, we want to look at the one used with `MBEDTLS_USE_PSA_CRYPTO`).

Note: this function only handles RSA PKCS#1 v1.5, so no need to worry about OAEP here (which is actually not supported in PK so far).

Similar: #5162 and #5274 are also creating a temporary PSA private key.

### PR DESCRIPTION
**Pull request by superna9999**
## Description
In `library/pk_wrap.c`, provide an implementation of `rsa_decrypt_wrap` using `psa_asymmetric_decrypt()` instead of `mbedtls_rsa_pkcs1_decrypt()` when `MBEDTLS_USE_PSA_CRYPTO` is enabled.

Resolves #5160 

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Todos
- [X] Implementation
- [X] Tests

## Steps to test or reproduce
test_suite_pk should run clean

### CODE REVIEW

File: include/mbedtls/psa_util.h
Comment (mprse) [inline include/mbedtls/psa_util.h:297]: Do we need to add this here? It seems we have the spacial issue to clean up translation functions (https://github.com/ARMmbed/mbedtls/issues/5514).

File: library/pk_wrap.c
Comment (mprse) [inline library/pk_wrap.c:143]: Same as for encryption. Please consider moving `RSA_PRV_DER_MAX_BYTES` to header file.

File: library/pk_wrap.c
Comment (mprse) [inline library/pk_wrap.c:201]: Same as for encryption. Status of psa_destroy_key() is ignored.

**Review by mprse — COMMENTED** [review]
Left a few comments similar to the once in encryption PR.

File: library/pk_wrap.c
Comment (mprse) [inline library/pk_wrap.c:191]: style: `olen);` -> `olen );`

File: include/mbedtls/psa_util.h
Comment (superna9999) [reply include/mbedtls/psa_util.h:297]: Yes, I have an implementation of such function to use here in #5519, this should probaly wait until #5519 is merged and re-use it

Comment (superna9999) [reply include/mbedtls/psa_util.h:297]: I'll move this as a special return case in `rsa_decrypt_wrap()` and clean this up in #5529

File: library/pk_wrap.c
Comment (superna9999) [reply library/pk_wrap.c]:
```
Naming patterns of macros: FAIL
    > include/mbedtls/pk.h:796: 'RSA_PUB_DER_MAX_BYTES' does not match the
      required pattern '^(MBEDTLS|PSA)_[0-9A-Z_]*[0-9A-Z]$'.
      |
  796 | #define RSA_PUB_DER_MAX_BYTES   ( 38 + 2 * MBEDTLS_MPI_MAX_SIZE )
      |         ^^^^^^^^^^^^^^^^^^^^^
    > include/mbedtls/pk.h:813: 'MPI_MAX_SIZE_2' does not match the required
      pattern '^(MBEDTLS|PSA)_[0-9A-Z_]*[0-9A-Z]$'.
      |
  813 | #define MPI_MAX_SIZE_2          ( MBEDTLS_MPI_MAX_SIZE / 2 + \
      |         ^^^^^^^^^^^^^^
    > include/mbedtls/pk.h:815: 'RSA_PRV_DER_MAX_BYTES' does not match the
      required pattern '^(MBEDTLS|PSA)_[0-9A-Z_]*[0-9A-Z]$'.
      |
  815 | #define RSA_PRV_DER_MAX_BYTES   ( 47 + 3 * MBEDTLS_MPI_MAX_SIZE \
      |         ^^^^^^^^^^^^^^^^^^^^^
    > include/mbedtls/pk.h:820: 'RSA_PUB_DER_MAX_BYTES' does not match the
      required pattern '^(MBEDTLS|PSA)_[0-9A-Z_]*[0-9A-Z]$'.
      |
  820 | #define RSA_PUB_DER_MAX_BYTES   0
      |         ^^^^^^^^^^^^^^^^^^^^^
    > include/mbedtls/pk.h:821: 'RSA_PRV_DER_MAX_BYTES' does not match the
      required pattern '^(MBEDTLS|PSA)_[0-9A-Z_]*[0-9A-Z]$'.
      |
  821 | #define RSA_PRV_DER_MAX_BYTES   0
```
Should I rename these ?

File: include/mbedtls/pk.h
Comment (mprse) [inline include/mbedtls/pk.h:779]: Seems like moving the macros to header file won't be so simple. CI is unhappy because of the macro names do not match pattern:

```
    > include/mbedtls/pk.h:796: 'RSA_PUB_DER_MAX_BYTES' does not match the
      required pattern '^(MBEDTLS|PSA)_[0-9A-Z_]*[0-9A-Z]$'
```

It would require to change the names of macros to `MBEDTLS_xxx` and adapt the existing code.

File: library/pk_wrap.c
Comment (mprse) [reply library/pk_wrap.c]: I missed this comment and left similar below.
I think yes. The changes should impact only one source file.

File: include/mbedtls/psa_util.h
Comment (mprse) [reply include/mbedtls/psa_util.h]: I have checked the https://github.com/ARMmbed/mbedtls/pull/5529 and the PR and I think that this one has better chances to go first. Translation function will probably require additional discussion and translation required here is pretty straight forward  `PSA_ERROR_INVALID_PADDING -> MBEDTLS_ERR_RSA_INVALID_PADDING`. When this is merged probably https://github.com/ARMmbed/mbedtls/pull/5529 will have to be rebased.

Sorry for changing my mind, but it seems that this way will be easier.

File: library/pk_wrap.c
Comment (mpg) [reply library/pk_wrap.c]: Yes, sorry I forgot to mention because it's so deeply ingrained: all public symbols must be prefixed with `MBEDTLS_` or `mbedtls_` (depending on the type of symbol). Usually this is followed by the name of the "module" that the symbol belongs to  (typically the name of the header file it's declared it); this is not always true for old things, but we should try to follow that rule for new things). For private symbols (static functions, etc) we can omit the `MBEDTLS_` prefix.

So, ideally this should be renamed to start with `MBEDTLS_PK_` since it's declared in `pk.h`.

Comment (superna9999) [reply library/pk_wrap.c]: Will be done in #5581 instead of currently pushed solution, will need to rebase when merged

File: include/mbedtls/pk.h
Comment (superna9999) [reply include/mbedtls/pk.h:779]: Will be done in #5581 instead of currently pushed solution, will need to rebase when merged

**Comment by superna9999** [comment]
I rebased against `development` after merge of #5581 and dropped the reverted commits moved to #5581.

**Review by mprse — APPROVED** [review]
LGTM

**Comment by superna9999** [comment]
I rebased against `development` since merge of #5529 and fixed usage of new PSA to mbedtls PK error mapping functions

**Review by mprse — APPROVED** [review]

**Review by mpg — CHANGES_REQUESTED** [review]
Looks pretty good to me, just left a couple suggestions for improvements.

File: library/pk_wrap.c
Comment (superna9999) [reply library/pk_wrap.c:229]: Ack, I'll do the same for the other similar PRs

File: library/pk_wrap.c
Comment (superna9999) [reply library/pk_wrap.c:235]: Sometimes I over-engineer stuff...

**Review by mpg — APPROVED** [review]
Thanks for addressing my feedback, looks all good to me now.

**Comment by mpg** [comment]
Regarding CI results: the internal CI fully passed, and the only failure on the Open CI is and instance of #5279 so it can be ignored as far as the present PR is concerned.

**Review by mprse — APPROVED** [review]

**Pull request merged**
